components/supervisor/pkg/serverapi/publicapi.go

```diff
@@ -0,0 +1,365 @@
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+)
+
+type APIInterface interface {
+	GetWorkspace(ctx context.Context, id string) (res *gitpod.WorkspaceInfo, err error)
+	GetToken(ctx context.Context, query *gitpod.GetTokenSearchOptions) (res *gitpod.Token, err error)
+	OpenPort(ctx context.Context, workspaceID string, port *gitpod.WorkspaceInstancePort) (res *gitpod.WorkspaceInstancePort, err error)
+	InstanceUpdates(ctx context.Context, instanceID string, workspaceID string) (<-chan *gitpod.WorkspaceInstance, error)
+
+	// Remove this and use segment client directly
+	TrackEvent(ctx context.Context, event *gitpod.RemoteTrackMessage) (err error)
+
+	// Metrics
+	RegisterMetrics(registry *prometheus.Registry) error
+}
+
+const (
+	// KindGitpod marks tokens that provide access to the Gitpod server API.
+	KindGitpod = "gitpod"
+)
+
+type ServiceConfig struct {
+	Host              string
+	Endpoint          string
+	InstanceID        string
+	WorkspaceID       string
+	SupervisorVersion string
+}
+
+type Service struct {
+	cfg         *ServiceConfig
+	experiments experiments.Client
+
+	token   string
+	ownerID string
+
+	lastServerInstance *gitpod.WorkspaceInstance
+
+	// gitpodService server API
+	gitpodService gitpod.APIInterface
+	// publicAPIConn public API publicAPIConn
+	publicAPIConn    *grpc.ClientConn
+	publicApiMetrics *grpc_prometheus.ClientMetrics
+
+	previousUsingPublicAPI bool
+}
+
+var _ APIInterface = (*Service)(nil)
+
+func NewServerApiService(ctx context.Context, cfg *ServiceConfig, tknsrv api.TokenServiceServer) APIInterface {
+	tknres, err := tknsrv.GetToken(context.Background(), &api.GetTokenRequest{
+		Kind: KindGitpod,
+		Host: cfg.Host,
+		Scope: []string{
+			"function:getToken",
+			"function:openPort",
+			"function:trackEvent",   // do we need this?
+			"function:getWorkspace", // do we need this?
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot get token for Gitpod API")
+		return nil
+	}
+	// server api
+	gitpodService, err := gitpod.ConnectToServer(cfg.Endpoint, gitpod.ConnectToServerOpts{
+		Token: tknres.Token,
+		Log:   log.Log,
+		ExtraHeaders: map[string]string{
+			"User-Agent":              "gitpod/supervisor",
+			"X-Workspace-Instance-Id": cfg.InstanceID,
+			"X-Client-Version":        cfg.SupervisorVersion,
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot connect to Gitpod API")
+		return nil
+	}
+
+	service := &Service{
+		token:         tknres.Token,
+		gitpodService: gitpodService,
+		cfg:           cfg,
+		experiments: experiments.NewCustomClient(configcat.Config{
+			SDKKey:       "gitpod",
+			BaseURL:      fmt.Sprintf("%s%s", os.Getenv("GITPOD_HOST"), "/configcat"),
+			PollInterval: time.Minute * 1,
+			HTTPTimeout:  3 * time.Second,
+		}),
```

Comment:
Let's bring this into the `experiments` package. I think we can re-use the existing `experiments.NewClient` which checks if the ENV var is set, and based on that we either use our proxy endpoint, or we use the upstream.